flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionReader.java

```diff
@@ -0,0 +1,191 @@
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.ResultSubpartition.BufferAndBacklog;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Iterator;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The reader (read view) of a BoundedBlockingSubpartition.
+ */
+class BoundedBlockingSubpartitionReader implements ResultSubpartitionView {
+
+	/** The result subpartition that we read. */
+	private final BoundedBlockingSubpartition parent;
+
+	/** Further byte buffers, to handle cases where there is more data than fits into
+	 * one mapped byte buffer (2GB = Integer.MAX_VALUE). */
+	private final Iterator<ByteBuffer> furtherData;
+
+	/** The reader/decoder to the memory mapped region with the data we currently read from.
+	 * Max 2GB large. Further regions may be in the {@link #furtherData} field.
+	 * Null once the reader empty or disposed.*/
+	@Nullable
+	private BoundedBlockingSubpartitionMemory.Reader data;
+
+	/** The next buffer (look ahead). Null once the data is depleted or reader is disposed. */
+	@Nullable
+	private Buffer nextBuffer;
+
+	/** The remaining number of data buffers (not events) in the result. */
+	private int bufferBacklog;
+
+	/** Flag whether this reader is released. Atomic, to avoid double release. */
+	private boolean isReleased;
+
+	/**
+	 * Convenience constructor that takes a single buffer.
+	 */
+	BoundedBlockingSubpartitionReader(
+			BoundedBlockingSubpartition parent,
+			ByteBuffer data,
+			int numBuffers) throws IOException {
+		this(parent, Collections.singleton(data), numBuffers);
+	}
+
+	/**
+	 * The general constructor that takes a queue of buffers containing the data.
+	 */
+	BoundedBlockingSubpartitionReader(
+			BoundedBlockingSubpartition parent,
+			Iterable<ByteBuffer> data,
+			int numBuffers) throws IOException {
+
+		checkArgument(numBuffers >= 0);
+
+		this.parent = checkNotNull(parent);
+		this.bufferBacklog = numBuffers;
+
+		this.furtherData = data.iterator();
+		checkArgument(furtherData.hasNext());
+
+		this.data = new BoundedBlockingSubpartitionMemory.Reader(furtherData.next());
+		this.nextBuffer = fillNextBuffer();
+	}
+
+	@Nullable
+	@Override
+	public BufferAndBacklog getNextBuffer() throws IOException {
+		final Buffer current = nextBuffer; // copy reference to stack
+
+		if (current == null) {
+			// as per contract, we must return null when the reader is empty,
+			// but also in case the reader is disposed (rather than throwing an exception)
+			return null;
+		}
+		if (current.isBuffer()) {
+			bufferBacklog--;
+		}
+
+		Buffer next = fillNextBuffer();
+		nextBuffer = next;
+
+		return new BufferAndBacklog(
+				current,
+				next != null,
+				bufferBacklog,
+				next != null && !next.isBuffer());
```

Comment:
Use `nextBufferIsEvent()` to replace `next != null && !next.isBuffer()` here?

Reply:
Similar as above, heap references versus stack references. Not a critical optimization, but seems to also not hurt here.